### PR TITLE
Coproc stuff base v1

### PR DESCRIPTION
**Pull request by andr2000**

### CODE REVIEW

**Comment by aanisov** [comment]
Also please change the platform code file naming to simple gx6xxx

**Comment by andr2000** [comment]
@aanisov done

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+#define RGXFW_CR_IRQ_CLEAR                            RGX_CR_META_SP_MSLVIRQSTATUS
+#define RGXFW_CR_IRQ_CLEAR_MASK                       RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_CLRMSK
+#endif
+
+static int gx6xxx_read(struct vcpu *v, mmio_info_t *info,
+                       register_t *r, void *priv)
+{
+    struct mmio *mmio = priv;
+    struct vcoproc_rw_context ctx;
+    static int start = 1;
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: bool please

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    struct gx6xxx_info *info = (struct gx6xxx_info *)coproc->priv;
+    uint32_t irq_status;
+    unsigned long flags;
+
+    spin_lock_irqsave(&coproc->vcoprocs_lock, flags);
+    irq_status = readl(info->reg_irq_status);
+    if (irq_status & RGXFW_CR_IRQ_STATUS_EVENT_EN)
+    {
+        struct vcoproc_instance *vcoproc;
+        struct vgx6xxx_info *vinfo;
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: struct vcoproc_instance *vcoproc = info->curr;
struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)vcoproc->priv;

File: xen/arch/arm/domain_build.c [hunk 3]
```diff
@@ -1332,7 +1332,8 @@ static int handle_node(struct domain *d, struct kernel_info *kinfo,
         return make_timer_node(d, kinfo->fdt, node);
 
 #ifdef CONFIG_HAS_COPROC
-    if ( device_get_class(node) == DEVICE_COPROC )
+    if ( (device_get_class(node) == DEVICE_COPROC) &&
```
Comment (aanisov) [inline xen/arch/arm/domain_build.c]: Is dt_device_used_by() the only way to realize that the coproc is not going to be shared?
What about xen,coproc property?

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    unsigned long flags;
+
+    spin_lock_irqsave(&curr->coproc->vcoprocs_lock, flags);
+    spin_unlock_irqrestore(&curr->coproc->vcoprocs_lock, flags);
+    return 0;
+}
+
+static int gx6xxx_ctx_switch_to(struct vcoproc_instance *next)
+{
+    struct gx6xxx_info *info = (struct gx6xxx_info *)next->coproc->priv;
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Is it possible for next to be NULL here?

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    struct gx6xxx_info *info;
+    char *reg_base;
+    int ret;
+
+    coproc = coproc_alloc(pdev, &gx6xxx_vcoproc_ops);
+    if ( IS_ERR_OR_NULL(coproc) )
+        return PTR_ERR(coproc);
+
+    if ( (coproc->num_irqs != 1) || (coproc->num_mmios != 1) )
+    {
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Please add corresponding #define-s and use them whenever possible.

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+static int gx6xxx_dt_probe(struct platform_device *pdev)
+{
+    struct coproc_device *coproc;
+    struct device *dev = &pdev->dev;
+    struct gx6xxx_info *info;
+    char *reg_base;
+    int ret;
+
+    coproc = coproc_alloc(pdev, &gx6xxx_vcoproc_ops);
+    if ( IS_ERR_OR_NULL(coproc) )
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: please dev_err here too

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        start = 0;
+        spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+        vcoproc_sheduler_vcoproc_wake(ctx.coproc->sched, ctx.vcoproc);
+        spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    }
+    vaddr = (uint32_t *)((char *)mmio->base + ctx.offset);
+    *r = readl(vaddr);
+    spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+    return 1;
+}
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Too many spin_lock_irqsave/spin_unlock_irqrestore here. You can optimize in a such way:
...
goto out;
...
out: 
   spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
   return 1;
}

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
+        vinfo->irq_status = r;
+        spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+        return 1;
+    }
+    vaddr = (uint32_t *)((char *)mmio->base + ctx.offset);
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Do we need vaddr/cast here? Can't we write in a such way:
writel(r, mmio->base + ctx.offset); ?

File: xen/arch/arm/coproc/coproc.c [hunk 9]
```diff
@@ -42,23 +43,45 @@ s_time_t vcoproc_context_switch(struct vcoproc_instance *curr,
...
+        wait_time = coproc->ops->ctx_switch_from(curr);
+
+        if ( wait_time == 0 )
+        {
+            if (curr->state == VCOPROC_RUNNING)
+                curr->state = VCOPROC_WAITING;
+            else
+                curr->state = VCOPROC_SLEEPING;
+        }
+        if ( wait_time )
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: maybe else here? or another check (assert) for wait_time<0?

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    struct vcoproc_rw_context ctx;
+    static int start = 1;
+    uint32_t *vaddr;
+    unsigned long flags;
+
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: ctx.vcoproc might be NULL (inside vcoproc_get_rw_context() we call coproc_get_vcoproc(d, ctx->coproc) to fill it)

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+DT_DEVICE_END
+
+/*
+ * Local variables:
+ * mode: C
+ * c-file-style: "BSD"
+ * c-basic-offset: 4
+ * indent-tabs-mode: nil
+ * End:
+ */
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: General comments:
1. There are places where you don't follow XEN code-style.
2. Always check for NULL before dereferencing passed to function pointers. Especially when function is global and may be called from different places in XEN.

File: xen/arch/arm/coproc/coproc.c [hunk 12]
```diff
@@ -96,6 +119,92 @@ static struct coproc_device *coproc_find_by_path(const char *path)
...
+                                  struct vcoproc_instance *vcoproc)
+{
+    struct coproc_device *coproc;
+
+    if ( !vcoproc )
+        return;
+
+    coproc = vcoproc->coproc;
+    coproc->ops->vcoproc_deinit(d, vcoproc);
+    spin_lock(&coproc->vcoprocs_lock);
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: I'm just curious shouldn't we remove vcoproc from the list prior to deinit?

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+    reg_base = (char *)coproc->mmios[0].base;
+    info->reg_irq_status = (uint32_t *)(reg_base + RGXFW_CR_IRQ_STATUS);
+    info->reg_irq_clear = (uint32_t *)(reg_base + RGXFW_CR_IRQ_CLEAR);
+
+    ret = request_irq(coproc->irqs[0], IRQF_SHARED,
+                      gx6xxx_irq_handler, "GPU GX6xxx irq", coproc);
+    if ( ret )
+    {
+        dev_err(dev, "failed to request irq (%u)\n", coproc->irqs[0]);
+        goto out_release_irqs;
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Why do you jump to release_irq if you couldn't request it?

File: xen/arch/arm/coproc/plat/gx6xxx.c [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+    struct mmio *mmio = priv;
+    struct vcoproc_rw_context ctx;
+    uint32_t *vaddr;
+    unsigned long flags;
+
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: Some question as above -> ctx.vcoproc might be NULL

File: xen/arch/arm/coproc/coproc.c [hunk 15]
```diff
@@ -199,12 +308,137 @@ bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
...
+    num_irqs = 0;
+    while ( (res = platform_get_resource(pdev, IORESOURCE_IRQ, num_irqs)) )
+        num_irqs++;
+
+    if ( !num_irqs )
+    {
+        printk("Failed to find at least one irq for \"%s\"\n",
+               dev_path(coproc->dev));
+        ret = -ENODEV;
+        goto out_free_irqs;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: Is it realy an issue to have no irqs on coproc?

File: xen/arch/arm/coproc/coproc.h [hunk 16]
```diff
@@ -66,30 +67,33 @@ struct coproc_device {
...
     struct vcoproc_scheduler *sched;
 };
 
 /* coproc callback functions */
-struct vcoproc_ops {
+struct coproc_ops {
     /* callback to perform initialization for the vcoproc instance */
-    struct vcoproc_instance *(*vcoproc_init)(struct domain *,
-                                             struct coproc_device *);
+    int (*vcoproc_init)(struct domain *, struct coproc_device *,
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: Why do you need struct coproc_device here? It is delivered with struct vcoproc_instance

File: xen/arch/arm/coproc/coproc.c [hunk 17]
```diff
@@ -96,6 +119,92 @@ static struct coproc_device *coproc_find_by_path(const char *path)
...
+               dev_path(coproc->dev));
+        return ERR_PTR(-ENOMEM);
+    }
+
+    vcoproc->coproc = coproc;
+    vcoproc->domain = d;
+    vcoproc->state = VCOPROC_UNKNOWN;
+    spin_lock_init(&vcoproc->lock);
+
+    ret = coproc->ops->vcoproc_init(d, coproc, vcoproc);
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: Why do you need a coproc here? You already put it into the vcoproc few lines above.

File: xen/arch/arm/coproc/coproc.h [hunk 18]
```diff
@@ -66,30 +67,33 @@ struct coproc_device {
...
-    bool_t (*vcoproc_is_created)(struct domain *, struct coproc_device *);
     /* callback to perform context switch from the running vcoproc instance */
     s_time_t (*ctx_switch_from)(struct vcoproc_instance *);
     /* callback to perform context switch to the waiting vcoproc instance */
     int (*ctx_switch_to)(struct vcoproc_instance *);
 };
 
+/* vcoproc read/write operation context */
+struct vcoproc_rw_context {
+    struct coproc_device *coproc;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: Couldn't be coproc acquired from vcoproc?

Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: Why do you need to keep coproc here? You always can get coproc from vcoproc.

File: xen/arch/arm/coproc/coproc.h [hunk 19]
```diff
@@ -141,6 +150,16 @@ int coproc_release_vcoprocs(struct domain *);
 
 #define dev_path(dev) dt_node_full_name(dev_to_dt(dev))
 
+static inline void vcoproc_get_rw_context(struct domain *d, struct mmio *mmio,
+                                          mmio_info_t *info,
+                                          struct vcoproc_rw_context *ctx)
+{
+    ctx->coproc = mmio->coproc;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: Well, it really should be checked why mmio holds coproc, not vcoproc

Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: I'm going to make a issue for the topic

File: xen/arch/arm/coproc/coproc.c [hunk 20]
```diff
@@ -199,12 +308,137 @@ bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
...
+        coproc->irqs[i] = irq;
+    }
+    coproc->num_irqs = num_irqs;
+
+    INIT_LIST_HEAD(&coproc->vcoprocs);
+    spin_lock_init(&coproc->vcoprocs_lock);
+    coproc->ops = ops;
+
+    return coproc;
+
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: The rollback actions below and their order remind me coproc_release(). So, can we optimize a little bit in a such way:
1. keep only one label out here where we call coproc_release()
2. Modify coproc_release() to check everything for NULL before releasing.
???

File: xen/arch/arm/coproc/coproc.h (hunk 19 above)
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: #20 is created and assigned to me. I'l start it tomorrow.

File: xen/arch/arm/coproc/plat/common.h [hunk 21]
```diff
@@ -24,40 +24,7 @@
...
-    printk(lvl "coproc: %s: " fmt, dt_node_full_name(dev_to_dt(dev)), ## __VA_ARGS__)
-
-#define dev_dbg(dev, fmt, ...) dev_print(dev, XENLOG_DEBUG, fmt, ## __VA_ARGS__)
-#define dev_notice(dev, fmt, ...) dev_print(dev, XENLOG_INFO, fmt, ## __VA_ARGS__)
-#define dev_warn(dev, fmt, ...) dev_print(dev, XENLOG_WARNING, fmt, ## __VA_ARGS__)
-#define dev_err(dev, fmt, ...) dev_print(dev, XENLOG_ERR, fmt, ## __VA_ARGS__)
-
-#define dev_err_ratelimited(dev, fmt, ...) \
-    dev_print(dev, XENLOG_ERR, fmt, ## __VA_ARGS__)
+struct coproc_device;
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/common.h]: Why do you still need `struct coproc_device` here?

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 22]
```diff
@@ -32,46 +32,15 @@ static s_time_t coproc_wait_time = MILLISECS(500);
 
 #define DT_MATCH_COPROC_XXX DT_MATCH_COMPATIBLE("vendor_xxx,coproc_xxx")
 
-static struct vcoproc_instance *coproc_xxx_get_vcoproc(struct domain *d,
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: Do you still need vmap.h in this file?

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 23]
```diff
@@ -121,96 +85,41 @@ static const struct mmio_handler_ops vcoproc_xxx_mmio_handler = {
...
-    if ( wait_time == 0 )
-    {
-        if (curr->state == VCOPROC_RUNNING)
-            curr->state = VCOPROC_WAITING;
-        else
-            curr->state = VCOPROC_SLEEPING;
-    }
-
-    return wait_time;
+    /* random for now */
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: if it seems to be a reference code, please put a comment inside, what should be done here

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 24]
```diff
@@ -121,96 +85,41 @@ static const struct mmio_handler_ops vcoproc_xxx_mmio_handler = {
...
 static int vcoproc_xxx_ctx_switch_to(struct vcoproc_instance *next)
 {
-    if ( !next )
-        return 0;
-
-    ASSERT(next->state == VCOPROC_WAITING);
-
-    next->state = VCOPROC_RUNNING;
-
+    /* nothing to do */
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: if it seems to be a reference code, please put a comment inside, what should be done here

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 25]
```diff
@@ -121,96 +85,41 @@ static const struct mmio_handler_ops vcoproc_xxx_mmio_handler = {
...
-    list_del(&vcoproc_xxx->vcoproc_elem);
-    spin_unlock(&coproc_xxx->vcoprocs_lock);
-    xfree(vcoproc_xxx);
-}
-
-static bool_t coproc_xxx_vcoproc_is_created(struct domain *d,
-                                            struct coproc_device *coproc_xxx)
-{
-    return coproc_xxx_get_vcoproc(d, coproc_xxx) ? true : false;
+    /* nothing to do */
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: if it seems to be a reference code, please put a comment inside, what should be done here

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 26]
```diff
@@ -336,18 +168,7 @@ static int coproc_xxx_dt_probe(struct platform_device *pdev)
...
-        if ( !IS_ERR(coproc_xxx->mmios[i].base) )
-            iounmap(coproc_xxx->mmios[i].base);
-    }
-out_free_mmios:
-    xfree(coproc_xxx->mmios);
-    xfree(coproc_xxx);
-
+    coproc_release(coproc_xxx);
     return ret;
 }
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: Some my comments from gx6xxx.c are applicable for coproc_xxx too. Please have a look

File: xen/arch/arm/coproc/coproc.c [hunk 27]
```diff
@@ -199,12 +308,137 @@ bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
...
+            goto out_free_irqs;
+        }
+        coproc->irqs[i] = irq;
+    }
+    coproc->num_irqs = num_irqs;
+
+    INIT_LIST_HEAD(&coproc->vcoprocs);
+    spin_lock_init(&coproc->vcoprocs_lock);
+    coproc->ops = ops;
+
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: Please check ops for NULL. Buggy platform driver might pass it.

File: xen/arch/arm/coproc/coproc.c [hunk 28]
```diff
@@ -21,6 +21,7 @@
 #include <xen/err.h>
 #include <xen/guest_access.h>
 #include <xen/keyhandler.h>
+#include <xen/vmap.h>
 
 #include "coproc.h"
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: I know that we have 4 platform ops at the moment and all of them seem to be mandatory. But, I think that you should add check for particularly operation to be present before call it.
e.x. 
if (coproc->ops->ctx_switch_from)
   coproc->ops->ctx_switch_from(curr);

File: xen/arch/arm/coproc/coproc.c [hunk 29]
```diff
@@ -199,12 +308,137 @@ bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
...
 }
 
+struct coproc_device *coproc_alloc(struct platform_device *pdev,
+                                   const struct coproc_ops *ops)
+{
+    struct coproc_device *coproc;
+    struct device *dev = &pdev->dev;
+    struct resource *res;
+    int num_irqs, num_mmios, i, ret;
+
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: Should we check somewhere that coproc_alloc have been already called for this platform device? Nothing prevents platform driver to call this function twice.

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 1 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: done

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 2 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: done

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 4 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: no, there is a check outside: if ( likely(next) )

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 5 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: done

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 6 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: won't fix: coproc_alloc already prints error messages

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 7 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: done

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 8 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: we can, but "(char *)mmio->base + ctx.offset" then, b/c base is a void pointer

File: xen/arch/arm/coproc/coproc.c (hunk 9 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: won't fix: original logic preserved. such changes should be in a separate change-set. I'll open an issue for that

Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: #21

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 10 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: I'll put
`BUG_ON(ctx->vcoproc != NULL);`

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 11 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: 1. ok
2. won't fix

File: xen/arch/arm/coproc/coproc.c (hunk 12 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: +1

Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: #22

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 13 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: fixed

File: xen/arch/arm/coproc/plat/gx6xxx.c (hunk 14 above)
Comment (andr2000) [inline xen/arch/arm/coproc/plat/gx6xxx.c]: fixed, same as above

File: xen/arch/arm/coproc/coproc.c (hunk 17 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: original code preserved: #23

File: xen/arch/arm/coproc/coproc.c (hunk 15 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: @otyshchenko1 could you please comment?
#24

File: xen/arch/arm/coproc/coproc.c (hunk 20 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: Agree, will do in a separate task #26 + #25

File: xen/arch/arm/coproc/coproc.c (hunk 27 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: tentative

File: xen/arch/arm/coproc/coproc.c (hunk 29 above)
Comment (andr2000) [inline xen/arch/arm/coproc/coproc.c]: tentative

**Pull request closed without merging**